CRVResponse/src/CrvWaveformsGenerator_module.cc

```diff
@@ -187,76 +186,104 @@ namespace mu2e
+
+        //find the TDC time when the first charge occurs (adjusted for this FEB)
+        double firstChargeTime=chargeClusters[iCluster].times.front();
+        firstChargeTime-=1.0*_digitizationPeriod;
+        double TDCstartTimeAdjusted=ceil((firstChargeTime-TDC0timeAdjusted)/_digitizationPeriod) * _digitizationPeriod + TDC0timeAdjusted;
+
+        //first create the full waveform
+        std::vector<double> fullWaveform;
+        _makeCrvWaveforms->MakeWaveform(chargeClusters[iCluster].times, chargeClusters[iCluster].charges,
+                                        fullWaveform, TDCstartTimeAdjusted, _digitizationPeriod);
+        _makeCrvWaveforms->AddElectronicNoise(fullWaveform, _noise, _randGaussQ);
+
+        //break the waveform apart into short pieces (CrvDigiMC::NSamples) and apply the zero suppression
+        //don't digitize outside of digitizationStart and digitizationEnd
+        for(size_t i=0; i<fullWaveform.size(); ++i)
         {
-          //start new single waveform
-          double digiStartTime=TDCstartTimeAdjusted+i*_digitizationPeriod;
-          if(digiStartTime<digitizationStart) continue; //digis cannot start before the digitization interval
-          if(digiStartTime>digitizationEnd) continue; //digis cannot start after the digitization interval
-//          if(digiStartTime+(CrvDigiMC::NSamples-1)*_digitizationPeriod>digitizationEnd) continue; //digis cannot end after the digitization interval
-
-          //collect voltages
-          std::array<double,CrvDigiMC::NSamples> voltages;
-          for(size_t singleWaveformIndex=0; singleWaveformIndex<CrvDigiMC::NSamples; i++, singleWaveformIndex++)
+          if(SingleWaveformStart(fullWaveform, i)) //acts as a zero suppression
           {
-            if(i<fullWaveform.size() && TDCstartTimeAdjusted+i*_digitizationPeriod<=digitizationEnd) voltages[singleWaveformIndex]=fullWaveform[i];
-            else voltages[singleWaveformIndex]=0.0;  //so that all unused single waveform samples are set to zero
-          }
+            //start new single waveform
+            double digiStartTime=TDCstartTimeAdjusted+i*_digitizationPeriod;
+            if(digiStartTime<digitizationStart) continue; //digis cannot start before the digitization interval
+            if(digiStartTime>digitizationEnd) continue; //digis cannot start after the digitization interval
+//            if(digiStartTime+(CrvDigiMC::NSamples-1)*_digitizationPeriod>digitizationEnd) continue; //digis cannot end after the digitization interval
+
+            //collect voltages
+            std::array<double,CrvDigiMC::NSamples> voltages;
+            for(size_t singleWaveformIndex=0; singleWaveformIndex<CrvDigiMC::NSamples; ++i, ++singleWaveformIndex)
+            {
+              if(i<fullWaveform.size() && TDCstartTimeAdjusted+i*_digitizationPeriod<=digitizationEnd) voltages[singleWaveformIndex]=fullWaveform[i];
```

Comment:
isn't this test (<=digitizationEnd) redundant with the above line 214
is it intentional that you are incrementing the outer loop variable (i++)?  That is also getting incremented in the loop.  Did you mean instead (line 207) while(i < fullWaveform.size())?

Reply:
Line 221 is not redundant. Line 214 checks the beginning of the 8-sample CrvDigi to see whether it should even exist. Line 221 checks individual samples of the CrvDigis, and fills the remaining samples past digitizationEnd and/or past the end of the full waveform (that was filled based on the arriving charges) with zeros. Of course the question remains how the real electronics responds at the end of the digitization window.
The outer loop (line 207) scans the full waveform until it finds relevant pieces to store in the CrvDigis. The inner loop (line 219) continues the loop through this waveform and fills the CrvDigis. Both loops advance the vector index i of the full waveform.